compendium/ConfigurationAdmin/test/TestConfigurationAdminImpl.cpp

```diff
@@ -41,6 +41,40 @@ MATCHER_P(AnyMapEquals, value, "")
   cppmicroservices::any_value_to_string(valueString, value);
   return (argString.str() == valueString.str());
 }
+/* This is a ConfigurationListener that blocks on a mutex. It is used by the TestConfigurationAdminImpl
+ * VerifyNoUpdateDeadlock test to recreate the blocking condition that caused a deadlock for the 
+ * Update, UpdateIfDifferent and Remove Configuration methods
+ */
+struct BlockingConfigurationListener final
+  : public cppmicroservices::service::cm::ConfigurationListener
+{
+
+  BlockingConfigurationListener() = default;
+  BlockingConfigurationListener(const BlockingConfigurationListener&) = delete;
+  BlockingConfigurationListener(BlockingConfigurationListener&&) = delete;
+  BlockingConfigurationListener& operator=(
+    const BlockingConfigurationListener&) = delete;
+  BlockingConfigurationListener& operator=(BlockingConfigurationListener&&) =
+    delete;
+  ~BlockingConfigurationListener() = default;
+
+  /*
+   * configurationEvent is the method called by Configuration Admin whenever a 
+   * a configuraton object is updated or removed.
+   *
+   */
+  void configurationEvent(
+    const cppmicroservices::service::cm::ConfigurationEvent& event) noexcept
+  {
+    {
+      std::lock_guard<std::mutex> lg{ mutex_ };
+    }
```

Comment:
This isn't locking anything. Is there a reason we acquire a lock, let the lock go out of scope (releasing our locked access to shared data), and then calling `promise_.set_value()` in an unlocked manner?

Reply:
The combination of this lock and the std::future being saved in the incompleteFutures vector is what recreates the deadlock that we saw in the user code. Certainly not anything you would do in normal code but the easiest way to recreate it.